src/lib.rs

```diff
@@ -186,7 +186,7 @@ impl HeapSizeOf for Url {
 pub struct ParseOptions<'a> {
     base_url: Option<&'a Url>,
     encoding_override: encoding::EncodingOverride,
-    log_syntax_violation: Option<&'a Fn(&'static str)>,
+    violation_fn: ViolationFn<'a>,
```

Comment:
Please store an `Option<ViolationFn<`'a>>` and remove the `NoOp` variant.

Reply:
If I did that, it would complicate things.  Firstly, I couldn't directly implement the helper methods `call` and `call_if`, given the outer type is now Option.  Secondly it would complicate those methods even if I could.  See the comments around `call_if` below.

Reply:
I see. Why not make the methods on `Parser` instead of `ViolationFn`? This would let you use `Option<T>`, right?

Reply:
OK, for preview, in 95a7b4a I reinstate Option and move the necessary methods back to the Parser.  This change would reduce the [overall diff](/servo/rust-url/compare/master...dekellum:95a7b4a) for the feature, keeping more status-quo with the parser.   However it spreads around the `if let Some(vfn) ...` and the `*_if` function can't be used outside of the `Parser` impl. `Input` already uses `ViolationFn::call`.  If the `*_if` variant is ever needed elsewhere this will prove a liability. 

I personally still think the current PR is better without this change, but if you prefer 95a7b4a, I will include it in this branch and PR.